AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m

```diff
@@ -300,28 +300,28 @@ - (void)flushQueueForTokenArray:(NSArray<MSAuthTokenValidityInfo *> *)tokenArray
                         if (logArray.count > 0) {
                           MSLogContainer *container = [[MSLogContainer alloc] initWithBatchId:batchId andLogs:logArray];
                           [self sendLogContainer:container withAuthTokenFromArray:tokenArray atIndex:tokenIndex];
-                        } else {
-
-                          // No logs available with given params.
-                          if (tokenIndex == 0 && tokenArray[tokenIndex].endTime != nil &&
-                              [self.storage countLogsBeforeDate:tokenArray[tokenIndex].endTime] == 0) {
-
-                            // Delete token from history if we don't have logs fitting it in DB.
-                            [[MSAuthTokenContext sharedInstance] removeAuthToken:tokenInfo.authToken];
-                          }
+                        }
 
-                          // Check to determine if the next index is within bounds.
-                          if (tokenIndex + 1 < tokenArray.count) {
+                        // No logs available with given params.
+                        else if (tokenIndex == 0 && tokenArray[tokenIndex].endTime != nil &&
+                                 [self.storage countLogsBeforeDate:tokenArray[tokenIndex].endTime] == 0) {
 
-                            // Iterate to next token in array.
-                            [self flushQueueForTokenArray:tokenArray withTokenIndex:tokenIndex + 1];
-                          }
+                          // Delete token from history if we don't have logs fitting it in DB.
+                          [[MSAuthTokenContext sharedInstance] removeAuthToken:tokenInfo.authToken];
                         }
                       }];
 
   // Flush again if there is another batch to send.
-  if (self.availableBatchFromStorage && !self.pendingBatchQueueFull) {
-    [self flushQueueForTokenArray:tokenArray withTokenIndex:tokenIndex];
+  if (!self.pendingBatchQueueFull) {
```

Comment:
This kind of seems like a band-aid for the problem.  Yes, we can just send more logs if they are present, but it seems like we should be able to determine which logs should be sent all at once... Maybe I'm misunderstanding, though?  It's two different ways to solve the same problem, but this seems like a reactive solution as opposed to a proactive one

Could you explain your logic for this particular solution @MatkovIvan ?

Reply:
@jwallra the issue is clear and this is not workaround or dirty fix or something like that.
Let me simplify this stuff (unwrap load function) and provide pseudo-code:

```
logArray = read_from_db(batchSizeLimit + 1)
if (logArray.count > batchSizeLimit) {
  logArray.removeLastObject();
  availableBatchFromStorage = YES;
}

// BEFORE:
noMoreLogs = logArray.count == 0;

// NOW:
noMoreLogs = !availableBatchFromStorage;
```